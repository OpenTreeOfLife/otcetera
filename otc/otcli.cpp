#include "otc/otcli.h"
#include <cstring>
#include <string>

///////////////////////////////////////////////////////////////
// pragmas are MTH mods to silence clang
#pragma clang diagnostic push
#pragma clang diagnostic ignored  "-Wglobal-constructors"
#pragma clang diagnostic ignored  "-Wexit-time-destructors"
INITIALIZE_EASYLOGGINGPP

static el::Configurations defaultConf;

#pragma clang diagnostic pop
///////////////////////////////////////////////////////////////

namespace otc {
bool debuggingOutputEnabled = false;
long ottIDBeingDebugged = -1;
OTCLI::OTCLI(const char *title,
          const char *descrip,
          const char *usage,
          bool quietExecution)
        :exitCode(0),
        verbose(false),
        currReadingDotTxtFile(false),
        blob(nullptr),
        titleStr(title),
        descriptionStr(descrip),
        usageStr(usage),
        out(std::cout),
        err(std::cerr) {
    defaultConf.setToDefault();
    if (quietExecution) {
        defaultConf.set(el::Level::Global, el::ConfigurationType::Enabled, "false");
    } else {
        defaultConf.set(el::Level::Trace, el::ConfigurationType::Enabled, "false");
        defaultConf.set(el::Level::Debug, el::ConfigurationType::Enabled, "false");
    }
    el::Loggers::reconfigureLogger("default", defaultConf);
}

void OTCLI::printHelp(std::ostream & outStream) {
    outStream << this->titleStr << ": " << this->descriptionStr << ".\n";
    outStream << "The most common usage is simply:\n";
    outStream << "    " << this->titleStr << " " << this->usageStr << "\n";
    outStream << "Standard command-line flags:\n";
    outStream << "    -h on the command line shows this help message\n";
    outStream << "    -fFILE treat each line of FILE as an arg\n";
    outStream << "    -q QUIET mode (all logging disabled)\n";
    outStream << "    -t TRACE level debugging (very noisy)\n";
    outStream << "    -v verbose\n";
    if (!clientDefFlagHelp.empty()) {
        outStream << "Exe-specific command-line flags:\n";
    }
    for (const auto & c : clientDefFlagHelp) {
        if (contains(clientDefArgNeeded, c.first)) {
            outStream << "    -" << c.first << "ARG " << c.second << '\n';
        } else {
            outStream << "    -" << c.first << ' ' << c.second << '\n';
        }
    }
}

void OTCLI::turnOnVerboseMode() {
    this->verbose = true;
    debuggingOutputEnabled = true;
    defaultConf.set(el::Level::Debug, el::ConfigurationType::Enabled, "true");
    el::Loggers::reconfigureLogger("default", defaultConf);
}

void OTCLI::turnOffVerboseMode() {
    this->verbose = false;
    debuggingOutputEnabled = false;
    defaultConf.set(el::Level::Debug, el::ConfigurationType::Enabled, "false");
    el::Loggers::reconfigureLogger("default", defaultConf);
}

bool OTCLI::handleFlag(const std::string & flagWithoutDash) {
    bool recursionNeeded = false;
    auto f = flagWithoutDash[0];
    if (clientDefFlagCallbacks.find(f) != clientDefFlagCallbacks.end()) {
        const auto cb = clientDefFlagCallbacks[f];
        try {
            const auto n = flagWithoutDash.substr(1);
            if (contains(clientDefArgNeeded, f) && n.empty()) {
                this->err << "Expecting an argument value after the  -" << f << " flag.\n";
                return false;
            }
            const auto rc = cb(*this, n);
            return rc;
        } catch (std::exception & x) {
            this->err << x.what() << '\n';
            return false;
        }
    }
    if (f == 'h') {
        this->printHelp(this->out);
        this->exitCode = 1;
        return false;
    } else if (f == 'v') {
        if (flagWithoutDash.length() > 1) {
            recursionNeeded = true;
        }
        turnOnVerboseMode();
    } else if (f == 't') {
        if (flagWithoutDash.length() > 1) {
            recursionNeeded = true;
        }
        this->verbose = true;
        defaultConf.set(el::Level::Trace, el::ConfigurationType::Enabled, "true");
        el::Loggers::reconfigureLogger("default", defaultConf);
    } else if (f == 'q') {
        if (flagWithoutDash.length() > 1) {
            recursionNeeded = true;
        }
        this->verbose = false;
        debuggingOutputEnabled = false;
        defaultConf.set(el::Level::Global, el::ConfigurationType::Enabled, "false");
        el::Loggers::reconfigureLogger("default", defaultConf);
    } else if (f == 'f') {
        if (flagWithoutDash.length() == 1) {
            this->err << "Expecting an argument value after the  -f flag.\n";
            return false;
        }
        extraArgs = readLinesOfFile(flagWithoutDash.substr(1));
    } else {
        std::cerr  << "Unrecognized flag: -" + flagWithoutDash << '\n';
        return false;
    }
    if (recursionNeeded) {
        return this->handleFlag(flagWithoutDash.substr(1));
    }
    return true;
}
bool OTCLI::parseArgs(int argc, char *argv[], std::vector<std::string> & args) {
    this->exitCode = 0;
    std::list<std::string> allArgs;
    for (int i = 1; i < argc; ++i) {
        const std::string filepath = argv[i];
        allArgs.push_back(filepath);
    }
    for (auto aIt = allArgs.begin(); aIt != allArgs.end(); ++aIt) {
        const auto & filepath = *aIt;
        const auto slen = filepath.length();
        if (slen > 1U && filepath[0] == '-') {
            extraArgs.clear();
            const std::string flagWithoutDash = filepath.substr(1);
            if (!this->handleFlag(flagWithoutDash)) {
                return false;
            }
            if (!extraArgs.empty()) {
                auto b = aIt;
                ++b;
                allArgs.insert(b, begin(extraArgs), end(extraArgs));
            }
        } else {
            const std::string filepathstr(filepath);
            args.push_back(filepathstr);
        }
    }
    return true;
}

bool OTCLI::isDotTxtFile(const std::string &fp) {
    const size_t fnl = fp.length();
    return (fp.substr(fnl - 4) == std::string(".txt"));
}

} // namespace otc
