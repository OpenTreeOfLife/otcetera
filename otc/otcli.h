#if !defined OTCLI_H
#define OTCLI_H
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>
#include "otc/otc_base_includes.h"
#include "otc/newick.h"
#include "otc/util.h"
#include "otc/tree_iter.h"
#include <boost/tokenizer.hpp>
#include <boost/program_options.hpp>
#include <memory>
namespace otc {
bool get_bool(const std::string& arg, const std::string& context);

class OTCLI {
    public:
        OTCLI(const char *title,
              const char *descrip,
              const char *usage,
              bool silentExecution=false);
        int exitCode;
        bool verbose;
        bool currReadingDotTxtFile;
        std::string currentFilename;
        std::string prefixForFiles;
        std::string currTmpFilepath;
        void * blob;

        void addFlag(char flag, const std::string & help, bool (*cb)(OTCLI &, const std::string &), bool argNeeded) {
            if (clientDefFlagHelp.find(flag) != clientDefFlagHelp.end()) {
                throw OTCError("Clashing flag assignment");
            }
            clientDefFlagHelp[flag] = help;
            clientDefFlagCallbacks[flag] = cb;
            if (argNeeded) {
                clientDefArgNeeded.insert(flag);
            }
        }
        bool handleFlag(const std::string & flagWithoutDash);
        bool parseArgs(int argc, char *argv[], std::vector<std::string> & args);
        void printHelp(std::ostream & outStream);
        
        /*int readFilepath(const std::string &fp,
                          ProcessedTreeValidationFunction func=0L,
                          void * blob=0L);
        // reads a NexSON v1.2 at filepath and returns a NxsSimpleTree with the associated treeID
        NxsSimpleTree * readTreeFromNexSONv_1_2(const std::string &filepath, const std::string & tree_id); */
        bool isDotTxtFile(const std::string &fp);
        auto getTitle() const {
            return this->titleStr;
        }
        ParsingRules & getParsingRules() {
            return parsingRules;
        }
        const ParsingRules & getParsingRules() const {
            return parsingRules;
        }
        void turnOnVerboseMode();
        void turnOffVerboseMode();
    private:
        std::string titleStr;
        std::string descriptionStr;
        std::string usageStr;
        std::map<char, std::string> clientDefFlagHelp;
        std::map<char, bool (*)(OTCLI &, const std::string &)> clientDefFlagCallbacks;
        std::set<char> clientDefArgNeeded;
        ParsingRules parsingRules;
        std::list<std::string> extraArgs;
    public:
        std::ostream & out;
        std::ostream & err;
};

template<typename T>
inline bool processTrees(const std::string& filename,
                         const ParsingRules& parsingRules,
                         std::function<bool (std::unique_ptr<T>)> treePtr)
{
    std::ifstream inp;
    if (!openUTF8File(filename, inp)) {
        throw OTCError("Could not open \"" + filename + "\"");
    }
    LOG(INFO) << "reading \"" << filename << "\"...";
    
    ConstStrPtr filenamePtr = ConstStrPtr(new std::string(filename));
    FilePosStruct pos(filenamePtr);
    unsigned treeNumInThisFile = 1;
    for (;;) {
        std::unique_ptr<T> nt = readNextNewick<T>(inp, pos, parsingRules);
        if (nt == nullptr) {
            break;
        }
        if (parsingRules.pruneUnrecognizedInputTips) {
            pruneTipsWithoutIds(*nt);
            if (nt->getRoot() == nullptr) {
                continue;
            }
        }
        std::string treeName = std::string("tree ") + std::to_string(treeNumInThisFile++);
        treeName.append(" from ");
        treeName.append(filepathToFilename(filename));
        nt->setName(treeName);
        const auto cbr = treePtr(std::move(nt));
        if (not cbr) {
            return false;
        }
    }
    return true;
}

template<typename Tree_t>
std::vector<std::unique_ptr<Tree_t>> get_trees(const std::vector<std::string>& filenames, const ParsingRules& rules)
{
    std::vector<std::unique_ptr<Tree_t>> trees;
    std::function<bool(std::unique_ptr<Tree_t>)> proc = [&](std::unique_ptr<Tree_t> t) {trees.push_back(std::move(t));return true;};
    for(auto& filename: filenames)
	otc::processTrees(filename,rules,proc);
    return trees;
}

template<typename Tree_t>
std::vector<std::unique_ptr<Tree_t>> get_trees(const std::string& filename, const ParsingRules& rules)
{
    const std::vector<std::string> filenameVec = {filename};
    return get_trees<Tree_t>(filenameVec, rules);
}

template<typename Tree_t>
std::unique_ptr<Tree_t> get_tree(const std::string& filename, const ParsingRules& rules)
{
    auto trees = get_trees<Tree_t>(filename, rules);
    return std::move(trees[0]);
}

template<typename Tree_t>
std::unique_ptr<Tree_t> get_tree(const std::string& filename)
{
    ParsingRules rules;
    rules.requireOttIds = false;
    return get_tree<Tree_t>(filename, rules);
}

template<typename T>
int treeProcessingMain(OTCLI & otCLI,
                          int argc,
                          char * argv[],
                          std::function<bool (OTCLI &, std::unique_ptr<T>)> treePtr,
                          int (*summarizePtr)(OTCLI &),
                          unsigned minNumTrees);

template<typename T>
inline int treeProcessingMain(OTCLI & otCLI,
                                 int argc,
                                 char * argv[],
                                 std::function<bool (OTCLI &, std::unique_ptr<T>)> treePtr,
                                 int (*summarizePtr)(OTCLI &),
                                 unsigned minNumTrees) {
    std::vector<std::string> filenameVec;
    if (!otCLI.parseArgs(argc, argv, filenameVec)) {
        otCLI.exitCode = 1;
        return otCLI.exitCode;
    }
    if (filenameVec.size() < minNumTrees) {
        otCLI.printHelp(otCLI.err);
        otCLI.err << otCLI.getTitle() << ": Expecting at least " << minNumTrees << " tree filepath(s).\n";
        otCLI.exitCode = 1;
        return otCLI.exitCode;
    }
    try {
        if (treePtr) {
            std::function<bool(std::unique_ptr<T>)> treePtr1 = 
                [&otCLI,&treePtr](std::unique_ptr<T> t)
                {return treePtr(otCLI,std::move(t));};
            
            for (const auto & filename : filenameVec) {
                otCLI.currentFilename = filepathToFilename(filename);
                const auto cbr = processTrees(filename, otCLI.getParsingRules(), treePtr1);
                if (not cbr) {
                    otCLI.exitCode = 2;
                    return otCLI.exitCode;
                }
            }
        }
        if (summarizePtr) {
            return summarizePtr(otCLI);
        }
    } catch (std::exception & x) {
        std::cerr << "ERROR. Exiting due to an exception:\n" << x.what() << std::endl;
        otCLI.exitCode = 3;
        return otCLI.exitCode;
    }
    return otCLI.exitCode;
}

template<typename Tree>
inline std::set<long> getAllOTTIds(const Tree& taxonomy) {
    std::set<long> o;
    for (auto nd : iter_node_const(taxonomy)) {
        if (nd->hasOttId()) {
            o.insert(nd->getOttId());
        }
    }
    return o;
}

template<>
inline std::set<long> getAllOTTIds(const TreeMappedWithSplits &taxonomy) {
    return taxonomy.getRoot()->getData().desIds;
}
 
template<typename T>
class TaxonomyDependentTreeProcessor {
    public:
        using node_type = typename T::node_type;
        using node_data_type = typename T::node_type::data_type;
        using tree_data_type = typename T::data_type;
        using tree_type = T;

        std::unique_ptr<T> taxonomy;
        std::set<long> ottIds;

        virtual bool processTaxonomyTree(OTCLI & otCLI) {
            ottIds = getAllOTTIds(*taxonomy);
            if (not taxonomy->getRoot()->hasOttId())
                throw OTCError()<<"Taxonomy root does not have an OTT ID!";
            otCLI.getParsingRules().ottIdValidator = &ottIds;
            otCLI.getParsingRules().includeInternalNodesInDesIdSets = false;
            return true;
        }
        virtual bool processSourceTree(OTCLI & , std::unique_ptr<T> tree) {
            assert(tree != nullptr);
            assert(taxonomy != nullptr);
            return true;
        }
        virtual bool summarize(OTCLI &) {
            return true;
        }
        TaxonomyDependentTreeProcessor()
            :taxonomy(nullptr) {
        }
        virtual ~TaxonomyDependentTreeProcessor(){}
    
};

template<typename T>
inline bool taxDependentProcessNextTree(OTCLI & otCLI, std::unique_ptr<T> tree) {
    TaxonomyDependentTreeProcessor<T> * tdtp = static_cast<TaxonomyDependentTreeProcessor<T> *>(otCLI.blob);
    assert(tdtp != nullptr);
    assert(tree != nullptr);
    if (tdtp->taxonomy == nullptr) {
        tdtp->taxonomy = std::move(tree);
        return tdtp->processTaxonomyTree(otCLI);
    }
    return tdtp->processSourceTree(otCLI, std::move(tree));
}

template<typename T>
int taxDependentTreeProcessingMain(OTCLI & otCLI,
                                   int argc,
                                   char *argv[],
                                   TaxonomyDependentTreeProcessor<T> & proc,
                                   unsigned int numTrees,
                                   bool includeInternalNodesInDesIdSets) {
    assert(otCLI.blob == nullptr);
    otCLI.blob = static_cast<void *>(&proc);
    otCLI.getParsingRules().includeInternalNodesInDesIdSets = includeInternalNodesInDesIdSets;
    std::function<bool (OTCLI &, std::unique_ptr<T>)> pcb = taxDependentProcessNextTree<T>;
    auto rc = treeProcessingMain<T>(otCLI, argc, argv, pcb, nullptr, numTrees);
    if (rc == 0) {
        return (proc.summarize(otCLI) ? 0 : 1);
    }
    return rc;
}

inline bool get_bool(const std::string& arg, const std::string& context) {
    if (arg == "true" or arg == "yes" or arg == "True" or arg == "Yes") {
        return true;
    } else if (arg == "false" or arg == "no" or arg == "False" or arg == "No") {
        return false;
    }
    throw OTCError() << context << "'" << arg << "' is not a recognized boolean value.";
}

boost::program_options::options_description standard_options();

boost::program_options::variables_map cmd_line_set_logging(const boost::program_options::variables_map& vm);

std::vector<std::string> cmd_line_response_file_contents(const boost::program_options::variables_map& vm);

boost::program_options::variables_map parse_cmd_line_response_file(int argc, char* argv[],
                                                                   const std::string& message,
                                                                   boost::program_options::options_description visible_,
                                                                   boost::program_options::options_description invisible,
                                                                   boost::program_options::positional_options_description p);

boost::program_options::variables_map parse_cmd_line_standard(int argc, char* argv[],
                                                              const std::string& message,
                                                              boost::program_options::options_description visible_,
                                                              boost::program_options::options_description invisible,
                                                              boost::program_options::positional_options_description p);

} // namespace otc
#endif
