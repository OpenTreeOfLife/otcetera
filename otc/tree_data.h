#ifndef OTCETERA_TREE_DATA_H
#define OTCETERA_TREE_DATA_H
// Classes that can serve as the template args for trees and nodes
#include <map>
#include <set>
#include "otc/otc_base_includes.h"

namespace otc {
template<typename, typename> class RootedTree;
template<typename> class RootedTreeNode;

template<typename T>
class RTreeOttIDMapping {
	public:
		typedef RootedTreeNode<T> NodeType;
		std::map<long, NodeType *> ottIdToNode;
		NodeType * getNodeForOttId(long ottId) const {
			const auto it = ottIdToNode.find(ottId);
			return (it == ottIdToNode.end() ? nullptr : it->second);
		}
};

class RTSplits {
	public:
		std::set<long> desIds;
};

} // namespace otc
#endif
