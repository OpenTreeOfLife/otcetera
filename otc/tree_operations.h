#ifndef OTCETERA_TREE_OPERATIONS_H
#define OTCETERA_TREE_OPERATIONS_H
// Functions that operate on trees - may include iteration
//	over trees (contrast w/tree_util.h)
// Depends on: tree.h tree_util.h tree_iter.h 
// Depended on by: tools

#include "otc/otc_base_includes.h"
#include "otc/tree_iter.h"

namespace otc {

template<typename T, typename U>
unsigned int countPolytomies(const RootedTree<T, U> & tree);


template<typename T, typename U>
unsigned int countPolytomies(const RootedTree<T, U> & tree) {
	unsigned int n = 0U;
	for (auto node : ConstPostorderInternalNode<T, U>{tree}) {
		if (node->getOutDegree() > 2) {
			n += 1;
		}
	}
	return n;
}

template<typename T, typename U>
void fillDesIdSets(RootedTree<T, U> & tree) {
	// assumes OttId is set for each tip
	for (auto node : PostorderInternalNode<T, U>(tree)) {
		std::set<long> & desIds = node->getData().desIds;
		if (node->IsTip()) {
			desIds.insert(node->getOttId());
		} else {
			for (auto child : ChildIterator<T, U>(*node)) {
				std::set<long> & cDesIds = child->getData().desIds;
				desIds.insert(cDesIds.begin(), cDesIds.end());
			}
		}
	}
}
} // namespace otc
#endif

