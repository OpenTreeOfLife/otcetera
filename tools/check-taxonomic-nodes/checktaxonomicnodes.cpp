#include "otc/otcli.h"
#include "otc/tree_operations.h"
#include "otc/tree_data.h"
using namespace otc;
typedef RTreeOttIDMapping<RTNodeNoData> RootedTreeForNodeType;
typedef RootedTree<RTNodeNoData, RootedTreeForNodeType> Tree_t;

bool processNextTree(OTCLI & otCLI, std::unique_ptr<Tree_t> tree);

struct CheckTaxonState {
	std::unique_ptr<Tree_t> toCheck;
	std::unique_ptr<Tree_t> taxonomy;
	int numErrors;

	CheckTaxonState()
		:toCheck(nullptr),
		 taxonomy(nullptr),
		 numErrors(0) {
		}
	void summarize(const OTCLI &otCLI) {

	}
};



inline bool processNextTree(OTCLI & otCLI, std::unique_ptr<Tree_t> tree) {
	CheckTaxonState * ctsp = static_cast<CheckTaxonState *>(otCLI.blob);
	assert(ctsp != nullptr);
	assert(tree != nullptr);
	if (ctsp->toCheck == nullptr) {
		ctsp->toCheck = std::move(tree);
	} else if (ctsp->taxonomy == nullptr) {
		ctsp->taxonomy = std::move(tree);
	} else {
		otCLI.err << "Expecting only 2 trees a tree to check and the taxonomy!\n";
		return false;
	}
	return true;
}

int main(int argc, char *argv[]) {
	OTCLI otCLI("otcchecktaxonomicnodes",
				"takes 2 newick file paths: to a tree with internal node names and a tree form of a taxonomy. Reports on any taxonomic names that do not map to the correct place in the first tree.",
				"some.tre taxonomy.tre");
	CheckTaxonState cts;
	otCLI.blob = static_cast<void *>(&cts);
	auto rc = treeProcessingMain<RTNodeNoData, RootedTreeForNodeType>(otCLI,
																	  argc,
																	  argv,
																	  processNextTree,
																	  nullptr);
	if (rc == 0) {
		cts.summarize(otCLI);
		return cts.numErrors;
	}
	return rc;
}

#if 0
/* use some globals, because I'm being lazy... */
NxsSimpleTree * gRefTree = 0;
NxsSimpleTree * gTaxonTree = 0;
std::map<long, const NxsSimpleNode *> gOttID2RefNode;
std::map<long, const NxsSimpleNode *> gOttID2TaxNode;
std::map<const NxsSimpleNode *, long> gTaxNode2ottID;
std::set<const NxsSimpleNode *> gSupportedNodes;

std::map<const NxsSimpleNode *, std::set<long> > gRefNdp2mrca;
std::map<const NxsSimpleNode *, std::set<long> > gTaxNdp2mrca;
set<long> gRefLeafSet;
set<long> gTaxLeafSet;
map<const NxsSimpleNode *, long> gRefNamedNodes;


void processRefTree(const NxsTaxaBlockAPI * tb, const NxsSimpleTree * tree) {
	std::vector<const NxsSimpleNode *> nodes =  tree->GetPreorderTraversal();
	for (std::vector<const NxsSimpleNode *>::const_reverse_iterator nIt = nodes.rbegin(); nIt != nodes.rend(); ++nIt) {
		const NxsSimpleNode * nd = *nIt;
		long ottID = getOTTIndex(tb, *nd);
		if (nd->IsTip()) {
			const unsigned ind = nd->GetTaxonIndex();
			assert(ind < tb->GetNumTaxonLabels());
			const std::string tn = tb->GetTaxonLabel(ind);
			assert(ottID >= 0);
			gRefNdp2mrca[nd].insert(ottID);
			gRefLeafSet.insert(ottID);
		} else {
			useChildrenToFillMRCASet(nd, gRefNdp2mrca);
			if (ottID > 0) {
				gRefNamedNodes[nd] = ottID;
			}
		}
		if (ottID >= 0) {
			assert(gOttID2RefNode.find(ottID) == gOttID2RefNode.end());
			gOttID2RefNode[ottID] = nd;
		}
	}
}

void processTaxonomyTree(const NxsTaxaBlockAPI * tb, const NxsSimpleTree * tree) {
	std::vector<const NxsSimpleNode *> nodes =  tree->GetPreorderTraversal();
	for (std::vector<const NxsSimpleNode *>::const_reverse_iterator nIt = nodes.rbegin(); nIt != nodes.rend(); ++nIt) {
		const NxsSimpleNode * nd = *nIt;
		long ottID = getOTTIndex(tb, **nIt);
		assert(ottID >= 0);
		assert(gOttID2TaxNode.find(ottID) == gOttID2TaxNode.end());
		if (nd->IsTip()) {
			assert(gOttID2RefNode.find(ottID) != gOttID2RefNode.end());
			gTaxLeafSet.insert(ottID);
			gTaxNdp2mrca[nd].insert(ottID);
		} else {
			useChildrenToFillMRCASet(nd, gTaxNdp2mrca);
		}
		gOttID2TaxNode[ottID] = *nIt;
		gTaxNode2ottID[*nIt] = ottID;
	}
	for (std::map<long, const NxsSimpleNode *>::const_iterator nit = gOttID2RefNode.begin(); nit != gOttID2RefNode.end(); ++nit) {
		assert(gOttID2TaxNode.find(nit->first) != gOttID2TaxNode.end());
	}
}

bool isProperSubset(const set<long> & small, const set<long> & big) {
	if (big.size() <= small.size()) {
		return false;
	}
	for (set<long>::const_iterator rIt = small.begin(); rIt != small.end(); ++rIt) {
		if (big.find(*rIt) == big.end()) {
			return false;
		}
	}
	return true;
}

bool doCheckEquivalent(std::ostream &out, long ottID, const NxsSimpleNode * snode, std::map<const NxsSimpleNode *, std::set<long> > & srcLookup,
										  const NxsSimpleNode * tnode, std::map<const NxsSimpleNode *, std::set<long> > & taxLookup,
										  bool topLevel, bool climbSynth, bool climbTax) {
	std::map<const NxsSimpleNode *, std::set<long> >::const_iterator streeLSIt = srcLookup.find(snode);
	assert(streeLSIt != srcLookup.end());
	std::map<const NxsSimpleNode *, std::set<long> >::const_iterator taxtreeLSIt = taxLookup.find(tnode);
	assert(taxtreeLSIt != taxLookup.end());
	const std::set<long> & streeMRCA = streeLSIt->second;
	const std::set<long> & taxtreeMRCA = taxtreeLSIt->second;
	if (streeMRCA != taxtreeMRCA) {
		if (topLevel) {
			out << "ottID " << ottID << " incorrect:\n";
			writeOttSetDiff(out, "    ", streeMRCA, "synth", taxtreeMRCA, "taxonomy");
		}
		if (climbSynth && isProperSubset(streeMRCA, taxtreeMRCA)) {
			return doCheckEquivalent(out, ottID, snode->GetEdgeToParent().GetParent(), srcLookup, tnode, taxLookup, false, true, false);
		} else if (climbTax && isProperSubset(taxtreeMRCA, streeMRCA)) {
			return doCheckEquivalent(out, ottID, snode, srcLookup, tnode->GetEdgeToParent().GetParent(), taxLookup, false, false, true);
		} else {
			return false;
		}
	} else if (!topLevel) {
		out << "        Found identical leaf sets for the synthetic tree \"" << snode->GetName() << "\" and the taxonomic node \"" << tnode->GetName() << "\".\n";
	}
	return true;
}

void summarize(std::ostream & out) {
	for (map<const NxsSimpleNode *, long>::const_iterator rnit = gRefNamedNodes.begin(); rnit != gRefNamedNodes.end(); ++rnit) {
		const NxsSimpleNode * nd = rnit->first;
		const long ottID = rnit->second;
		std::map<long, const NxsSimpleNode *>::const_iterator tID2nd = gOttID2TaxNode.find(ottID);
		assert(tID2nd != gOttID2TaxNode.end());
		const NxsSimpleNode *taxNd = tID2nd->second;
		if (!doCheckEquivalent(out, ottID, nd, gRefNdp2mrca, taxNd, gTaxNdp2mrca, true, true, true)) {
			out << "        Could not find this set of leaves in the synth \"" << nd->GetName() <<"\" in any taxonomic node.\n";
		}
	}
	if (gTaxLeafSet != gRefLeafSet) {
		writeOttSetDiff(out, "", gRefLeafSet, "synth", gTaxLeafSet, "taxonomy");
	}
}

bool newTreeHook(NxsFullTreeDescription &ftd, void * arg, NxsTreesBlock *treesB) {
	const NxsTaxaBlockAPI * taxa = treesB->GetTaxaBlockPtr();
	NxsSimpleTree * nst = new NxsSimpleTree(ftd, 0.0, 0, true);
	if (gRefTree == 0) {
		gRefTree = nst;
		processRefTree(taxa, nst);
	} else if (gTaxonTree == 0) {
		gTaxonTree = nst;
		processTaxonomyTree(taxa, nst);
	} else {
		const char * msg = "Exepting only 2 files: the tree file, and then the taxonomy\n";
		std::cerr << msg;
		throw NxsException(msg);
	}
	if (gRefTree != nst && gTaxonTree != nst) {
		delete nst;
	}
	return false;
}

int main(int argc, char *argv[]) {
	std::vector<std::string> args;
	if (!gOTCli.parseArgs(argc, argv, args)) {
		return 1;
	}
	if (args.size() != 2) {
		cerr << "Expecting a tree file and taxonomy tree.\n";
		return 1;
	}
	try{
		gOTCli.readFilepath(args[0], newTreeHook);
		gOTCli.readFilepath(args[1], newTreeHook);
	} catch (NxsException &x) {
		std::cerr << x.what() << "\n";
		return 1;
	}
	if (gOTCli.exitCode == 0) {
		summarize(std::cout);
	}
	return gOTCli.exitCode;
}

#endif
