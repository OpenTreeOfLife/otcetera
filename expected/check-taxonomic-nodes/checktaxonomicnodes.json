[
  {
    "invocation" : ["otc-check-taxonomic-nodes", "<INFILELIST>"],
    "infile_list": ["3genus-mislabeled.tre", "3genus-taxonomy.tre"],
    "expected": "checktaxonomicnodes-expected"
  }
]
